src/Microsoft.Health.Dicom.Core/Features/CustomTag/ReindexJob.cs

```diff
@@ -0,0 +1,79 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Dicom.Core.Features.Model;
+using Microsoft.Health.Dicom.Core.Models;
+
+namespace Microsoft.Health.Dicom.Core.Features.CustomTag
+{
+    public class ReindexJob : IReindexJob
+    {
+        private readonly ICustomTagStore _customTagStore;
+        private readonly IInstanceIndexer _instanceIndexer;
+
+        // Start from 1 for now, we can consider bigger number for better performance later.
+        private const int Top = 1;
+
+        public ReindexJob(ICustomTagStore customTagStore, IInstanceIndexer instanceIndexer)
+        {
+            EnsureArg.IsNotNull(customTagStore, nameof(customTagStore));
+            EnsureArg.IsNotNull(instanceIndexer, nameof(instanceIndexer));
+            _customTagStore = customTagStore;
+            _instanceIndexer = instanceIndexer;
+        }
+
+        public async Task ReindexAsync(IEnumerable<CustomTagStoreEntry> customTagStoreEntries, long endWatermark, CancellationToken cancellationToken)
+        {
+            // TODO: this is transient solution for main scenario to go through. There are several things need to be done when move onto job framework.
```

Comment:
Lets have a conversation with Brandon on what is realisticaly covered by Job framework. Lets not assume things.
There are couple of things missing in this solution
1. The work needs to happen in batches, this won't scale from get go
2. Log everything, long running jobs are bound to fail, this helps to debug issues in prod easily
3. No matter where the code get executed on persisting the intermediate state, code needs to be return CustomTag.